### ISSUE TITLE
What is the Difference Between a Batch and an Epoch in a Neural Network?

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#449]
What is the Difference Between a Batch and an Epoch in a Neural Network?<br>
<br>
https://ift.tt/2LqTCS4<br>
<br>
Jason Brownlee<br>
<br>
<p>Stochastic gradient descent is a learning algorithm that has a number of hyperparameters.</p>
<p>Two hyperparameters that often confuse beginners are the batch size and number of epochs. They are both integer values and seem to do the same thing.</p>
<p>In this post, you will discover the difference between batches and epochs in stochastic gradient descent.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>Stochastic gradient descent is an iterative learning algorithm that uses a training dataset to update a model.</li>
<li>The batch size is a hyperparameter of gradient descent that controls the number of training samples to work through before the model&rsquo;s internal parameters are updated.</li>
<li>The number of epochs is a hyperparameter of gradient descent that controls the number of complete passes through the training dataset.</li>
</ul>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/07/What-is-the-Difference-Between-a-Batch-and-an-Epoch-in-a-Neural-Network.jpg" alt="What is the Difference Between a Batch and an Epoch in a Neural Network?" width="640" class="wp-image-5820" height="346">
<p class="wp-caption-text">What is the Difference Between a Batch and an Epoch in a Neural Network?<br>
Photo by <a href="https://www.flickr.com/photos/grazza123/9754438586/">Graham Cook</a>, some rights reserved.</p>
</div>
<h2>Overview</h2>
<p>This post is divided into five parts; they are:</p>
<ol>
<li>Stochastic Gradient Descent</li>
<li>What Is a Sample?</li>
<li>What Is a Batch?</li>
<li>What Is an Epoch?</li>
<li>What Is the Difference Between Batch and Epoch?</li>
</ol>
<h2>Stochastic Gradient Descent</h2>
<p>Stochastic Gradient Descent, or SGD for short, is an optimization algorithm used to train machine learning algorithms, most notably artificial neural networks used in deep learning.</p>
<p>The job of the algorithm is to find a set of internal model parameters that perform well against some performance measure such as logarithmic loss or mean squared error.</p>
<p>Optimization is a type of searching process and you can think of this search as learning. The optimization algorithm is called &ldquo;<em>gradient descent</em>&ldquo;, where &ldquo;<em>gradient</em>&rdquo; refers to the calculation of an error gradient or slope of error and &ldquo;descent&rdquo; refers to the moving down along that slope towards some minimum level of error.</p>
<p>The algorithm is iterative. This means that the search process occurs over multiple discrete steps, each step hopefully slightly improving the model parameters.</p>
<p>Each step involves using the model with the current set of internal parameters to make predictions on some samples, comparing the predictions to the real expected outcomes, calculating the error, and using the error to update the internal model parameters.</p>
<p>This update procedure is different for different algorithms, but in the case of artificial neural networks, the <a href="https://en.wikipedia.org/wiki/Backpropagation">backpropagation</a> update algorithm is used.</p>
<p>Before we dive into batches and epochs, let&rsquo;s take a look at what we mean by sample.</p>
<p>Learn more about gradient descent here:</p>
<ul>
<li><a href="https://machinelearningmastery.com/gradient-descent-for-machine-learning/">Gradient Descent For Machine Learning</a></li>
</ul>
<h2>What Is a Sample?</h2>
<p>A sample is a single row of data.</p>
<p>It contains inputs that are fed into the algorithm and an output that is used to compare to the prediction and calculate an error.</p>
<p>A training dataset is comprised of many rows of data, e.g. many samples. A sample may also be called an instance, an observation, an input vector, or a feature vector.</p>
<p>Now that we know what a sample is, let&rsquo;s define a batch.</p>
<h2>What Is a Batch?</h2>
<p>The batch size is a hyperparameter that defines the number of samples to work through before updating the internal model parameters.</p>
<p>Think of a batch as a for-loop iterating over one or more samples and making predictions. At the end of the batch, the predictions are compared to the expected output variables and an error is calculated. From this error, the update algorithm is used to improve the model, e.g. move down along the error gradient.</p>
<p>A training dataset can be divided into one or more batches.</p>
<p>When all training samples are used to create one batch, the learning algorithm is called batch gradient descent. When the batch is the size of one sample, the learning algorithm is called stochastic gradient descent. When the batch size is more than one sample and less than the size of the training dataset, the learning algorithm is called mini-batch gradient descent.</p>
<ul>
<li>
<strong>Batch Gradient Descent</strong>. Batch Size = Size of Training Set</li>
<li>
<strong>Stochastic Gradient Descent</strong>. Batch Size = 1</li>
<li>
<strong>Mini-Batch Gradient Descent</strong>. 1 &lt; Batch Size &lt; Size of Training Set</li>
</ul>
<p>In the case of mini-batch gradient descent, popular batch sizes include 32, 64, and 128 samples. You may see these values used in models in the literature and in tutorials.</p>
<p><strong>What if the dataset does not divide evenly by the batch size?</strong></p>
<p>This can and does happen often when training a model. It simply means that the final batch has fewer samples than the other batches.</p>
<p>Alternately, you can remove some samples from the dataset or change the batch size such that the number of samples in the dataset does divide evenly by the batch size.</p>
<p>For more on the differences between these variations of gradient descent, see the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/gentle-introduction-mini-batch-gradient-descent-configure-batch-size/">A Gentle Introduction to Mini-Batch Gradient Descent and How to Configure Batch Size</a></li>
</ul>
<p>A batch involves an update to the model using samples; next, let&rsquo;s look at an epoch.</p>
<h2>What Is an Epoch?</h2>
<p>The number of epochs is a hyperparameter that defines the number times that the learning algorithm will work through the entire training dataset.</p>
<p>One epoch means that each sample in the training dataset has had an opportunity to update the internal model parameters. An epoch is comprised of one or more batches. For example, as above, an epoch that has one batch is called the batch gradient descent learning algorithm.</p>
<p>You can think of a for-loop over the number of epochs where each loop proceeds over the training dataset. Within this for-loop is another nested for-loop that iterates over each batch of samples, where one batch has the specified &ldquo;batch size&rdquo; number of samples.</p>
<p>The number of epochs is traditionally large, often hundreds or thousands, allowing the learning algorithm to run until the error from the model has been sufficiently minimized. You may see examples of the number of epochs in the literature and in tutorials set to 10, 100, 500, 1000, and larger.</p>
<p>It is common to create line plots that show epochs along the x-axis as time and the error or skill of the model on the y-axis. These plots are sometimes called learning curves. These plots can help to diagnose whether the model has over learned, under learned, or is suitably fit to the training dataset.</p>
<p>For more on diagnostics via learning curves with LSTM networks, see the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/diagnose-overfitting-underfitting-lstm-models/">How to Diagnose Overfitting and Underfitting of LSTM Models</a></li>
</ul>
<p>In case it is still not clear, let&rsquo;s look at the differences between batches and epochs.</p>
<h2>What Is the Difference Between Batch and Epoch?</h2>
<p>The batch size is a number of samples processed before the model is updated.</p>
<p>The number of epochs is the number of complete passes through the training dataset.</p>
<p>The size of a batch must be more than or equal to one and less than or equal to the number of samples in the training dataset.</p>
<p>The number of epochs can be set to an integer value between one and infinity. You can run the algorithm for as long as you like and even stop it using other criteria besides a fixed number of epochs, such as a change (or lack of change) in model error over time.</p>
<p>They are both integer values and they are both hyperparameters for the learning algorithm, e.g. parameters for the learning process, not internal model parameters found by the learning process.</p>
<p>You must specify the batch size and number of epochs for a learning algorithm.</p>
<p>There are no magic rules for how to configure these parameters. You must try different values and see what works best for your problem.</p>
<h3>Worked Example</h3>
<p>Finally, let&rsquo;s make this concrete with a small example.</p>
<p>Assume you have a dataset with 200 samples (rows of data) and you choose a batch size of 5 and 1,000 epochs.</p>
<p>This means that the dataset will be divided into 40 batches, each with five samples. The model weights will be updated after each batch of five samples.</p>
<p>This also means that one epoch will involve 40 batches or 40 updates to the model.</p>
<p>With 1,000 epochs, the model will be exposed to or pass through the whole dataset 1,000 times. That is a total of 40,000 batches during the entire training process.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<ul>
<li><a href="https://machinelearningmastery.com/gradient-descent-for-machine-learning/">Gradient Descent For Machine Learning</a></li>
<li><a href="https://machinelearningmastery.com/gentle-introduction-mini-batch-gradient-descent-configure-batch-size/">A Gentle Introduction to Mini-Batch Gradient Descent and How to Configure Batch Size</a></li>
<li><a href="https://machinelearningmastery.com/diagnose-overfitting-underfitting-lstm-models/">How to Diagnose Overfitting and Underfitting of LSTM Models</a></li>
<li><a href="https://en.wikipedia.org/wiki/Stochastic_gradient_descent">Stochastic gradient descent on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Backpropagation">Backpropagation on Wikipedia</a></li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered the difference between batches and epochs in stochastic gradient descent.</p>
<p>Specifically, you learned:</p>
<ul>
<li>Stochastic gradient descent is an iterative learning algorithm that uses a training dataset to update a model.</li>
<li>The batch size is a hyperparameter of gradient descent that controls the number of training samples to work through before the model&rsquo;s internal parameters are updated.</li>
<li>The number of epochs is a hyperparameter of gradient descent that controls the number of complete passes through the training dataset.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/difference-between-a-batch-and-an-epoch/">What is the Difference Between a Batch and an Epoch in a Neural Network?</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
July 20, 2018 at 03:12AM
